pkg/ruby/check_upgrade.go

```diff
@@ -0,0 +1,199 @@
+		}
+
+		cached, err = os.Create(cachedPath)
+		if err != nil {
+			return "", fmt.Errorf("failed to create cache file: %w", err)
+		}
+
+		_, err = cached.Write(decodedContent)
+		if err != nil {
+			return "", fmt.Errorf("failed to write cache file: %w", err)
+		}
+	}
+	defer cached.Close()
+	return cachedPath, nil
+}
+
+// checkVersionConstraint takes a gemspec and extracts the version constraints
+// and checks them against the RubyUpdateVersion. This function attempts to use
+// inline ruby to determine if the version is within constraints. If ruby is
+// not installed on the system then the terraform go-version package is used as
+// it has very similar version constraints. The inline ruby is the preferred
+// method in case something changes (highly unlikely).
+//
+// NOTE: If there is no required_ruby_version found in the gemspec then we
+//
+//	assume there is no limit.
+func (o *Options) checkVersionConstraint(gemspecFile string) error {
+	cached, err := os.Open(gemspecFile)
+	if err != nil {
+		return fmt.Errorf("opening gemspec file: %w", err)
+	}
+	versionLine := ""
+	scanner := bufio.NewScanner(cached)
+	for scanner.Scan() {
+		if strings.Contains(scanner.Text(), requiredRubyVersionKey) {
+			versionLine = scanner.Text()
+			break
+		}
+	}
+
+	// Could not find required_ruby_version in gemspec, assuming no constraint
+	if versionLine == "" {
+		return nil
+	}
+
+	pattern := `["']([^"']+)["']`
+	re := regexp.MustCompile(pattern)
+	groups := re.FindAllStringSubmatch(versionLine, -1)
+
+	versionConstraints := []string{}
+	for _, group := range groups {
+		versionConstraints = append(versionConstraints, group[1])
+	}
+
+	if len(versionConstraints) < 1 {
+		return fmt.Errorf("could not extract %s from gemspec", requiredRubyVersionKey)
+	}
+
+	rubyPath, err := exec.LookPath("ruby")
+	if err == nil { // Call ruby to do the comparison
+		// Need to construct a list of version constraints to pass
+		// to the ruby call. It should end up like this: ['>= 2.6', '< 4']
+		constraint := "["
+		for i, c := range versionConstraints {
+			constraint += fmt.Sprintf("'%s'", c)
+			if i != len(versionConstraints)-1 {
+				constraint += ", "
+			}
+		}
+		constraint += "]"
+
+		cmdArr := []string{
+			"-e",
+			fmt.Sprintf("Gem::Dependency.new('', %s).match?('', '%s') ? exit : abort('')", constraint, o.RubyUpdateVersion),
+		}
+		cmd := exec.Command(rubyPath, cmdArr...)
+		_, err := cmd.CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("ruby version comparison failed")
+		}
+	} else { // If ruby doesn't exist on the system do the version comparison ourselves
+		// Terraform has the same version comparison operators as ruby so we
+		// just use the terraform library to do the comparison
+		v1, err := version.NewVersion(o.RubyUpdateVersion)
+		if err != nil {
+			return fmt.Errorf("converting ruby update version: %w", err)
+		}
+
+		constraints, err := version.NewConstraint(strings.Join(versionConstraints, ", "))
```

Comment:
What's the difference between this and the preceding concatenation? Can you use the same result from before here?

Reply:
Unfortunately not, since one string is actually a valid array in ruby and the other is a function argument they are different